sabakan/generator.go

```diff
@@ -526,36 +522,27 @@ func (g *Generator) decreaseControlPlane() (*updateOp, error) {
 		return nil, nil
 	}
 
-	workers := make([]*Machine, len(g.workers))
-	for i, n := range g.workers {
-		workers[i] = g.machineMap[n.Address]
-	}
-
-	cps := make([]*Machine, len(g.controlPlanes))
-	for i, n := range g.controlPlanes {
-		cps[i] = g.machineMap[n.Address]
-	}
-
 	op := &updateOp{
-		name:    "decrease control plane",
-		workers: workers,
-		cps:     cps,
+		name: "decrease control plane",
 	}
+	g.nextControlPlanes = g.nodesToMachines(true)
+	g.nextWorkers = g.nodesToMachines(false)
 
-	for len(cps) != g.constraints.ControlPlaneCount {
+	for len(g.nextControlPlanes) != g.constraints.ControlPlaneCount {
 		var m *Machine
-		numCPs := op.countMachinesByRack(true)
-		m, cps = g.deselectMachine(true, cps, numCPs)
+		numCPs := g.countControlPlanesByRack()
+		m = g.deselectControlPlane(numCPs)
 
-		if g.constraints.MaximumWorkers == 0 || len(op.workers) < g.constraints.MaximumWorkers {
+		if g.constraints.MaximumWorkers == 0 || len(g.nextWorkers) < g.constraints.MaximumWorkers {
 			op.demoteControlPlane(m)
+			g.nextWorkers = append(g.nextWorkers, m)
 			continue
 		}
 
+		g.nextUnused = append(g.nextUnused, m)
```

Comment:
How should CKE work when the worker num exceeds MaximumWorkers due to demoting cp to a worker?
Should CKE remain the worker in the cluster?

Reply:
Doing so would remove a running worker node to keep satisfying `MaximumWorkers`.
It is not clear which is better.

We may improve the integration strategy not to reach `MaximumWorkers` to keep
a room for demoting control planes.